src/UserSpaceInstrumentation/AllocateInTracee.cpp

```diff
@@ -0,0 +1,193 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "AllocateInTracee.h"
+
+#include <absl/strings/numbers.h>
+#include <absl/strings/str_format.h>
+#include <absl/strings/str_split.h>
+#include <errno.h>
+#include <sys/mman.h>
+#include <sys/ptrace.h>
+#include <sys/wait.h>
+
+#include <cstdint>
+#include <string>
+#include <vector>
+
+#include "OrbitBase/Logging.h"
+#include "OrbitBase/ReadFileToString.h"
+#include "OrbitBase/SafeStrerror.h"
+#include "UserSpaceInstrumentation/RegisterState.h"
+
+namespace orbit_user_space_instrumentation {
+
+namespace {
+
+using orbit_base::ReadFileToString;
+
+// Returns the address range of the first executable memory region. In every case I encountered this
+// was the second line in the `maps` file corresponding to the code of the process we look at.
+// However we don't really care. So keeping it general and just search for a executable region is
+// probalby helping stability.
+[[nodiscard]] ErrorMessageOr<void> GetFirstExecutableMemoryRegion(pid_t pid, uint64_t* addr_start,
+                                                                  uint64_t* addr_end) {
+  auto result_read_maps = ReadFileToString(absl::StrFormat("/proc/%d/maps", pid));
+  if (result_read_maps.has_error()) {
+    return result_read_maps.error();
+  }
+  const std::vector<std::string> lines =
+      absl::StrSplit(result_read_maps.value(), '\n', absl::SkipEmpty());
+  for (const auto& line : lines) {
+    const std::vector<std::string> tokens = absl::StrSplit(line, ' ', absl::SkipEmpty());
+    if (tokens.size() < 2 || tokens[1].size() != 4 || tokens[1][2] != 'x') continue;
+    const std::vector<std::string> addresses = absl::StrSplit(tokens[0], '-');
+    if (addresses.size() != 2) continue;
+    *addr_start = std::stoull(addresses[0], nullptr, 16);
+    *addr_end = std::stoull(addresses[1], nullptr, 16);
+    return outcome::success();
+  }
+  return ErrorMessage(absl::StrFormat("Unable to locate executable memory area in pid: %d", pid));
+}
+
+// Write `bytes` into memory of thread `tid` starting from `address_start`.
+// Note that we write multiples of eight bytes at once. If length of `bytes` is not a multiple of
+// eight the remaining bytes are zeroed out.
+void WriteBytesIntoTraceesMemory(pid_t tid, uint64_t address_start,
+                                 const std::vector<uint8_t>& bytes) {
+  size_t pos = 0;
+  do {
+    // Pack 8 byte for writing into `data`.
+    uint64_t data = 0;
+    for (size_t i = 0; i < 8 && pos + i < bytes.size(); i++) {
+      uint64_t t = bytes[pos + i];
+      data |= t << (8 * i);
+    }
+    CHECK(ptrace(PTRACE_POKEDATA, tid, address_start + pos, data) != -1);
+    pos += 8;
```

Comment:
Is it overkill to have a constant and assign it `sizeof(data)` or `sizeof(uint64_t)` instead of using `8` directly?

Reply:
Pretty sure they are the same thing but it might document where the "8" is coming from.